crates/bin/pd/src/migrate/testnet72.rs

```diff
@@ -0,0 +1,206 @@
+            ..bsod
+        })
+    }
+
+    async fn translate_compact_block(
+        &self,
+        compact_block: pb::compact_block::v1::CompactBlock,
+    ) -> anyhow::Result<pb::compact_block::v1::CompactBlock> {
+        let mut swap_outputs = Vec::with_capacity(compact_block.swap_outputs.len());
+        for bsod in compact_block.swap_outputs {
+            swap_outputs.push(self.translate_bsod(bsod).await?);
+        }
+        Ok(pb::compact_block::v1::CompactBlock {
+            swap_outputs,
+            ..compact_block
+        })
+    }
+}
+
+/// Translate all of the BSODs inside dex storage to the new format.
+async fn translate_dex_storage(
+    ctx: Context,
+    delta: &mut StateDelta<Snapshot>,
+) -> anyhow::Result<()> {
+    let mut stream = delta.prefix_raw("dex/output/");
+    while let Some(r) = stream.next().await {
+        let (key, bsod_bytes) = r?;
+        let bsod = pb::dex::v1::BatchSwapOutputData::decode(bsod_bytes.as_slice())?;
+        let bsod = ctx.translate_bsod(bsod).await?;
+        delta.put_proto(key, bsod);
+    }
+    Ok(())
+}
+
+/// Translate all of the compact block storage to hold the new BSOD data inside the compact blocks.
+async fn translate_compact_block_storage(
+    ctx: Context,
+    delta: &mut StateDelta<Snapshot>,
+) -> anyhow::Result<()> {
+    let mut stream = delta.prefix_raw("compactblock/");
+    while let Some(r) = stream.next().await {
+        let (key, compactblack_bytes): (String, Vec<u8>) = r?;
+        let block = pb::compact_block::v1::CompactBlock::decode(compactblack_bytes.as_slice())?;
+        let block = ctx.translate_compact_block(block).await?;
+        delta.put_proto(key, block);
+    }
+    Ok(())
+}
+
+/// Run the full migration, given an export path and a start time for genesis.
+pub async fn migrate(
+    path_to_export: PathBuf,
+    genesis_start: Option<tendermint::time::Time>,
+) -> anyhow::Result<()> {
+    let rocksdb_dir = path_to_export.join("rocksdb");
+    let storage = Storage::load(rocksdb_dir.clone(), SUBSTORE_PREFIXES.to_vec()).await?;
+    let export_state = storage.latest_snapshot();
+    let root_hash = export_state.root_hash().await.expect("can get root hash");
+    let pre_upgrade_root_hash: RootHash = root_hash.into();
+    let pre_upgrade_height = export_state
+        .get_block_height()
+        .await
+        .expect("can get block height");
+    let post_upgrade_height = pre_upgrade_height.wrapping_add(1);
+
+    let mut delta = StateDelta::new(export_state);
+    let (migration_duration, post_upgrade_root_hash) = {
+        let start_time = std::time::SystemTime::now();
+        let ctx = Context::new(storage.clone());
+
+        // Translate inside dex storage.
+        translate_dex_storage(ctx.clone(), &mut delta).await?;
+        // Translate inside compact block storage.
+        translate_compact_block_storage(ctx.clone(), &mut delta).await?;
+
+        delta.put_block_height(0u64);
```

Comment:
Why do we reset the block height here?

Reply:
This is so we can do an `InitChain` handshake with comet and allow it to initialize its internal view before block sync/proposals. The flow looks like this:

```
pd <-   info query   -- comet
pd --  app height=0  -> comet
pd <- init chain req -- comet
pd -- init chain resp -> comet
```

unfortunately, the CometBft spec seem to have been reorganized again, so it will take me a minute to fish out the link

Reply:
Interesting. So the idea is that setting it to `0` signals that the app is ready for genesis, then once `pd` sends the init chain message, it will get set to `h+1` as set in the genesis file?

Reply:
That's right, it will consume the genesis as a result of reading a zero app height, and then pull the one specified in the file (`initialHeight`)